gearbot_2_lib/src/datastore/mod.rs

```diff
@@ -0,0 +1,129 @@
+use sqlx::postgres::PgPoolOptions;
+use tracing::{info, trace};
+use twilight_model::id::GuildId;
+use crate::datastore::crypto::EncryptionKey;
+use crate::datastore::guild::{GuildConfig, DatabaseGuildInfo, GuildInfo, GuildConfigWrapper};
+
+mod crypto;
+pub mod guild;
+mod error;
+
+pub use error::DatastoreError;
+
+pub type DatastoreResult<T> = Result<T, DatastoreError>;
+type Transaction<'a> = SqlxTransaction<'a, Postgres>;
+
+
+pub struct Datastore {
+    master_encryption_key: EncryptionKey<'static>,
+    pub(crate) pool: PgPool,
+}
+
+impl Datastore {
+    pub async fn initialize() -> Result<Self, sqlx::Error> {
+        info!("Initializing datastore...");
+        let database_url = env::var("DATABASE_URL").expect("Missing DATABASE_URL!");
+        let encryption_key = env::var("ENCRYPTION_KEY").expect("Missing ENCRYPTION_KEY!");
+
+        let pool = PgPoolOptions::new()
+            .max_connections(env::var("POOL_CONNECTIONS").map(|val| val.parse::<u32>().expect("Pool connections value isn't a proper number")).unwrap_or_else(|_| 5))
+            .idle_timeout(Duration::from_secs(60))
+            .connect(&database_url).await?;
+
+        info!("Pool created, making sure the database is up to date.");
+        sqlx::migrate!("../migrations").run(&pool).await?;
+        info!("Database migrations complete!");
+
+        Ok(
+            Datastore {
+                master_encryption_key: EncryptionKey::construct_owned(encryption_key.as_bytes()),
+                pool,
+            }
+        )
+    }
+
+    /// get the config and encryption key for a guild. if none exists one will be created.
+    /// If there was a left_at attribute it is now cleared
+    pub async fn get_or_create_guild_info(&self,  guild_id: &GuildId) -> DatastoreResult<GuildInfo> {
+        let mut transaction = self.pool.begin().await?;
+        let info: Option<DatabaseGuildInfo> = query_as!(
+            DatabaseGuildInfo,
+            "UPDATE guild_config SET left_at=null where id=$1 RETURNING id, version, config, encryption_key",
+            guild_id.get() as i64)
+            .fetch_optional(&mut transaction).await?;
+        let info= if let Some(info) = info {
+            if !info.has_supported_config() {
+                return Err(DatastoreError::UnsupportedConfigVersion(info.version));
+            }
+
+           info.into_config_and_key()?
+        } else {
+            // none existed, make a new one
+            let info = self.setup_new_guild(guild_id, &mut transaction).await?;
+            info
+        };
+        transaction.commit().await?;
+
+        Ok(info)
+
+    }
+
+    /// create and persist a new config and encryption key for a guild
+    async fn setup_new_guild(&self, guild_id: &GuildId, transaction: &mut Transaction<'_>) -> DatastoreResult<GuildInfo> {
+        let config = GuildConfig::default().wrapped();
+        let raw_config = serde_json::to_value(&config)?;
```

Comment:
Do these `query` macros really not support anything except `serde_json::Value`?

Reply:
that's what it has mapped to the json data type. i'd need like a custom serializer thing for it otherwise